src/dataset/impl_dataset.rs

```diff
@@ -1113,11 +1123,11 @@ pub trait IntoTargets<T> {
     fn into(self) -> T;
 }
 
-impl<F, D: Data<Elem = F>> IntoTargets<ArrayBase<D, Ix2>> for ArrayBase<D, Ix1> {
-    fn into(self) -> ArrayBase<D, Ix2> {
-        self.insert_axis(Axis(1))
-    }
-}
+// impl<F, D: Data<Elem = F>> IntoTargets<ArrayBase<D, Ix2>> for ArrayBase<D, Ix1> {
```

Comment:
I don't think we need this impl anymore. It converts Array1d into Array2d in `Dataset`, which creates conflicting implementations in the tests.

Reply:
This change has a lot of consequences. A lot of 1d targets were implicitly converted to Array2d, which is expected for functions like `cross_validate` to work. For `DatasetBase<ArrayBase<D, Ix2>, ArrayBase<S, Ix2>>`, there are `cross_validate` and `cross_validate_multi`. IMHO, we should remove `cross_validate_multi` from the trait and leave `cross_validate`, then we implement `cross_validate` for `DatasetBase<ArrayBase<D, Ix2>, ArrayBase<S, Ix1>>`. WDYT?

Reply:
That whole trait should be removed. The fact that it has `into` as a method is a red flag, since that makes it a worse version of `Into`. From what I understand the problem you're having currently is that, because most `Dataset` methods were implemented on 2D targets, they won't the compile now that we have datasets with 1D targets. For methods that are applicable for both 1D and 2D targets (most methods are like this), place them in a generic impl block bound by `AsMultiTargets`, since `as_multi_targets` normalizes 1D and 2D targets into 2D arrays. Methods applicable only for 1D targets should be implemented for `AsSingleTarget` if possible; otherwise implement for `ArrayBase<S, Ix1>`. Methods applicable only for 2D targets should be implemented for `AsMultiTargets` if possible; otherwise implement for `ArrayBase<S, Ix2>`. This strategy should minimize caller breakage.

For `cross_validate` and `cross_validate_multi`, I'm pretty sure that `cross_validate` applies to both 1D and 2D targets, while `cross_validate_multi` applies only to 2D and works a bit differently. `multi` allows the evaluation function to differentiate between different targets, since its evaluation function takes `Array2` instead of `Array1`. However, it's possible to express the semantics of 2D `cross_validate` in terms of `cross_validate_multi`, so we don't really need 2 functions. All we really need is one generic `cross_validate` function that covers both 1D and 2D targets, and it'll behave like `cross_validate_multi` in the 2D case. This should be easily doable by making the evaluation closure generic on the target dimensionality.

Reply:
It is well noted. I'm left confused by the `FromTargetArray` trait. It is not implemented for Array1d. However, `bootstrap`, `bootstrap_samples`, etc. are implemented with the `FromTargetArray` and `AsMultiTargets` traits. I only see one implementation of `FromTargetArray` for `CountedTargets` but none for Array1d and Array2d. I don't know how to deal with it.

Reply:
`FromTargetArray` acts like the reverse of `AsMultipleTargets`. It takes a 2D target array and converts it back into the original "target type", such as `CountedTargets`. If implemented on `Array2` it converts the target array into itself (identity operation pretty much). If implemented on `CountedTargets` it wraps the target array into a `CountedTargets`. Typically, a function like `bootstrap` takes some targets, calls `as_multi_targets` on it, manipulates the 2D target array, then calls `from_target` to turn it back into the original target type. This doesn't work on 1D targets right now, since `FromTargetArray` is not implemented on `Array1`. The impl should turn the 2D target array (which we know only has one target) into a 1D target array. Something like:
```rust
impl<'a, L: Clone + 'a, S: Data<Elem = L>> FromTargetArray<'a, L> for ArrayBase<S, Ix1> {
    type Owned = Array1<L>;
    type View = ArrayView1<'a, L>;

    fn new_targets(targets: Array2<L>) -> Self::Owned {
        // something like targets.reshape(targets.nrows()).unwrap()
    }

    fn new_targets_view(targets: ArrayView2<'a, L>) -> Self::View {
        // same here
    }
}
```